### ISSUE TITLE
Can't set initial values of heavy widgets

### ISSUE THREAD

**Issue by Peter-Gr** [applegrew/django-select2#222]
I am having some trouble getting initial values set correctly in a Heavy Widget in django-select2

All works fine when using the light widget - but the heavy widgets (which otherwise get populated with data fine when being clicked on) don't retain their previous data when re-loading the form data.

The issue seems to be in django_select2/forms.py
```python
render_options(self, choices, selected_choices):
```
In my case selected choices correctly contains the key values of the previously selected items.

If I am reading the code right -
```python
choices = {(k, v) for k, v in choices if k in selected_choices}
```
looks through the passed in choices values and creates a list of those which are also present in selected_choices.

That would work fine if choices contained the full list of possible values (which it does at the application level).

From what I can see render_options is called by django_select2 - 
```python
render(self, name, value, attrs=None, choices=()):
```
That expects a choices parameter otherwise it uses the default of no choices - which appears to be what is happening.

This code is called from (in django 1.8.3) django/forms/forms.py line 593
```python
return force_text(widget.render(name, self.value(), attrs=attrs))
```
When it calls widget.render it doesn't pass in any choices - so the default empty choices is used in the django_select2 code.

Is the issue with the django_select2 code? Or is it expecting to be called from something other than that line in django 1.8.3?

**Comment by codingjoe** [comment]
Hi @Peter-Gr,

the heavy fields only render selected choices, to reduce load. That's why they're called heavy, to, for example, render foreign keys to large tables.

What do you mean by initial value? Do you actually mean a form initial? I actually don't know those values are included in the selected template. If not they must be included in the choices, otherwise it wont show. The latter makes sense I guess, the further should be investigated.

Do you have any code example to reproduce this behaviour?

Thanks,
Joe

**Comment by Peter-Gr** [comment]
Thanks for any help on this. Yep I am using the heavy control as I need the web-page to go off to the database to get a particular set of options based on certain criteria on the page. That side of it all works fine 

There are a number of form controls on the page set up as:

def CreateMultiChoiceField(label_text, choice_values):
    return MultipleChoiceField(label = label_text,
                                      choices = choice_values,
                                      required=False,
                                      widget=HeavySelect2MultipleWidget(attrs={'class':'form-control searchable-control',
                                                                          'data-placeholder' : 'All'},
                                                                        data_url = 'updateselections'))


On creation choice_values is pre-populated with the set of all possible values (20 or 30 values). This are subsequently not used, as the choices are populated through an ajax query, but it looks like they are used in the code noted above, with the noted issues.

The view method which displays the page is pretty standard:

def index(request):
    if request.method == 'POST':
        form = FilterForm(request.POST)
        
        if form.is_valid():
            return render_to_response('graphs/index.html',{'durationchart': GetChart(form.cleaned_data), 
                                                              'displaygraph' : True,
                                                              'form'         : form},
                              context_instance=RequestContext(request))
    else:
        form = FilterForm()
    
    return render_to_response('graphs/index.html',{'displaygraph' : False,
                                                      'form'         : form},
                              context_instance=RequestContext(request))

The part which isn't working is the following flow.

1 - User opens web page with blank form data.
2 - User selected some items in the multi-select-heavy widgets.
3 - User presses submit to process the form data.
4 - This causes the page to be re-displayed with the previously filled in form data at the top and a graph displayed at the bottom based on the data which was entered into the form,
5 - the re-displayed page should have all the form fields filled in with the data which they contained before the form was submitted, so that the user can tweak the form values if they want and then re-display the graph with different options selected.

This all works perfectly for the other controls on the form, and even for the Light widget, but the heavy widget is not populating the option tags like:

<option value="2">DUPLICATE</option>
<option value="1" selected="selected">NORMAL</option>
<option value="3">OUTAGE</option>

From what I can tell this is required in order to have the select field display some pre-selected values on page-load.

This logic is contained in the django_select2 code:

    def render_options(self, choices, selected_choices):
        """Render only selected options."""
        output = ['<option></option>' if not self.is_required else '']
        choices = {(k, v) for k, v in choices if k in selected_choices}
        selected_choices = {force_text(v) for v in selected_choices}
        for option_value, option_label in choices:
            output.append(self.render_option(selected_choices, option_value, option_label))
        return '\n'.join(output)

Due to the issue I noted above, it looks like choices is blank when this is called and thus no option values are rendered to the HTML....

**Comment by codingjoe** [comment]
Thanks for the detailed report. Do I get it right, that you're submitting the form, with a required field missing, and then previously selected fields are empty?

**Comment by Peter-Gr** [comment]
Hi, no, no fields are missing initially.
Lets say I have a form with two fields. A string field and a Heavy Widget field.
I open the form and for the string I fill in "My Name" and for the Heavy widget select items "Apple" and "Banana".
I then press submit.
When submitting the form, some action happens and the same form is re-loaded with the previously user-entered data displayed.
So when the form is re-loaded it should still contain. The text field "My Name" and the selected items "Apple" and "Banana"

The re-display of previously entered form data is handled by passing in 'form' in the code

```python
            return render_to_response('graphs/index.html',{'durationchart':                      GetChart(form.cleaned_data), 
                                                              'displaygraph' : True,
                                                              'form'         : form},
```

The re-display works correctly for all of the non-heavy widget fields - but the heavy widgets go back to being blank.

I believe this is due to choices being empty, as noted, in the call to render_options quoted above.

**Comment by codingjoe** [comment]
Oh, ok, well, are your selected choices in your choices? I doesn't seem so.

**Comment by Peter-Gr** [comment]
The selected choices are in the choices setup when creating the widget. The issue is that when Django  (in django 1.8.3)  django/forms/forms.py line 593 calls

```python
return force_text(widget.render(name, self.value(), attrs=attrs))
```

it does not pass in those choices, so iterating through them in the select2 code doesn't find them. I'll work on putting together a minimal set of code that illustrates the issue.

**Comment by codingjoe** [comment]
Thanks, I get the picture. I don't like that they even handle choices as an instance attribute and method attribute.

**Comment by codingjoe** [comment]
Please check if it works with the patch i wrote.

**Comment by Peter-Gr** [comment]
Thanks - I'll give it a try!

**Comment by Peter-Gr** [comment]
I got it to work - but needed one further change to the line:
<code python>
choices = chain(choices, self.choices)
</code>
That returns an object of type
choices =  itertools.chain object at 0x7fd55873ed10
Then the line
<code python>
choices = {(k, v) for k, v in choices if k in selected_choices}
</code>

fails as the 'in' operator doesn't appear to work on objects of type itertools.chain

I fixed it by changing the first line to 
<code python>
choices = list(chain(choices, self.choices))
</code>
You might have another preferred way of doing the same thing - but it needs something like this - also the same presumably in the matching change further down.
thanks for the fix!

**Comment by codingjoe** [comment]
@Peter-Gr that isn't true
```python
>>> from itertools import chain
>>> [i for i in chain([1,2,],[4,5])]
[1, 2, 4, 5]
>>> 1 in chain([1,2,],[4,5])
True
```

**Comment by Peter-Gr** [comment]
Ha - thats a funny thing. 

When I first did the testing I things were not working due to a mis-match in data types between the key variable types in my choices and the key variable types in my selected choices - while tracing through that I added some print statements to the code to check what was happening.

Once I fixed the key type mis-match I left the print statements in place..

It looks like iterating through the values in the iterchain in my debug code was moving a pointer in the iterchain object
<pre>
>>> from itertools import chain
>>> choices = chain([1,2],[3,4])
>>> [ i for i in choices]
[1, 2, 3, 4]
>>> [ i for i in choices]
[]
>>> 
</pre>

So given that - and re-testing then yes the fix works correctly as is - thanks

**Comment by codingjoe** [comment]
Woohoo! @Peter-Gr thanks for your great work debuging, testing and reporting it in the first place.

Released in 5.4.2

**Comment by roboslone** [comment]
Hi! Not sure if it's the right place, but I couldn't find any docs concerning initial value set for `ModelSelect2Widget`. Is there a way to set `request.user` as initial value? I wasn't able to do so from view.

**Comment by codingjoe** [comment]
Hi @roboslone,

you usually don't set default values on widgets but rather on forms. Therefore there is no different behavior if you use  the `ModelSelect2Widget`. You can use `initial_data` or `get_inital_data` on a `FormView` in Django. Let me know if you need more help finding the proper documentation.

Cheers,
Joe

**Comment by roboslone** [comment]
@codingjoe , thanks for reply, I found a mistake in my code.

**Comment by timothedelion** [comment]
Hi !

I think I face a quite similar problem with `ModelSelect2Widget`. But actually, the problem might come from the interaction with `django_filters.FilterView`

It correctly work to select items, and build the right query params in URL, but when I reload the page, the ModelSelect2Widget does not display the previously selected item.

I put more details on my issue on SO : https://stackoverflow.com/questions/57480475/field-initial-data-does-not-display-in-a-modelselect2widget

Sorry if it is not as bright and precise as @Peter-Gr issue, I'm quite beginner.

**Comment by codingjoe** [comment]
Hi @timothedelion thanks for reaching out. I am sure we can find a solution together. Could you do me a favor and open a new issue, since this seems to be a different problem and I'd like the whole issue history to be here and not on StackOverflow. Best -Joe

**Comment by kunambi** [comment]
Hi @codingjoe 

Sorry to necro this thread, but I seem to have the same issue which apparently was fixed in patch `5.4.2` - we're currently using `django-select2==8.4.3`. We're currently facing the same issue (initial values are not rendered). 

We are also using django-select2 together with django-taggit.

I'll share some of the code:
```python
# views.py
from .models import Media

class DashboardMediaEditView(UpdateView):
    template_name = "ecard/dashboard/media_edit.html"
    queryset = Media.objects.all()
    form_class = DashboardMediaCreateForm


# forms.py
from django import forms
from django_select2 import forms as s2forms
from taggit.models import Tag

class MediaTagChoices(s2forms.ModelSelect2TagWidget):
    queryset = Tag.objects.all()
    search_fields = ["name__icontains"]

    @property
    def empty_label(self):
        return _("Search for, or create tags")

    def value_from_datadict(self, data, files, name):
        values = set(super().value_from_datadict(data, files, name))
        names = self.queryset.filter(**{"name__in": list(values)}).values_list(
            "name", flat=True
        )
        cleaned_values = list(names)

        for val in values - set(list(names)):
            cleaned_values.append(self.queryset.create(name=val).name)

        return cleaned_values


class DashboardMediaCreateForm(forms.ModelForm):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        if instance := kwargs.get("instance"):
            self.fields["tags"].queryset = Tag.objects.all()
            self.fields["tags"].widget = MediaTagChoices(
                to_field_name="name",
                model=Tag,
                initial=[tag.pk for tag in instance.tags.all()],
                choices=[(tag.pk, tag.name) for tag in instance.tags.all()],
                data_view="dashboard:mediatags",
                search_fields=["name__icontains"],
                attrs={
                    "debug": "true",
                    "data-minimum-input-length": 2,
                    "data-allow-clear": "false",
                    "data-placeholder": _("Tags"),
                },
            )

    def clean_tags(self):
        cleaned_values = []

        for data_tag in self.data.getlist("tags"):
            if data_tag.isdigit() and (tag := Tag.objects.filter(pk=data_tag).first()):
                cleaned_values.append(tag.name)
            else:
                cleaned_values.append(data_tag)

        return cleaned_values

    class Meta:
        model = Media
        labels = {"is_hidden": _("Hide it from managers")}
        widgets = {
            "file": forms.FileInput(attrs={"accept": "image/*, video/*"}),
        }
        fields = [
            "file",
            "tags",
            "is_hidden",
        ]
```

_Searching_ for tags, _creating_ new tags, and _saving_ to the database works perfectly fine! However, when the user clicks the Media object in order to edit it (perhaps wants to add another tag, or toggle the `is_hidden` field) the tags-field is _empty_. And if the user clicks the save-button, the Media object will be updated with _no_ tags (empty list).

This is the rendered HTML, so you can see that the saved tags are clearly there (`option` tags inside the `select` tag):
```html
<label>
  <select name="tags" lang="en" data-minimum-input-length="2" data-theme="default" data-allow-clear="false" data-placeholder="Tags" data-tags="true" data-token-separators="[&quot;,&quot;, &quot; &quot;]" data-ajax--url="/dashboard/media/tags/" data-ajax--cache="true" data-ajax--type="GET" debug="true" aria-describedby="id_tags_helptext" id="id_tags" class="django-select2 django-select2-heavy select2-hidden-accessible" data-field_id="Ijc0ODMxZjJhLTBkYWItNDczZC04MTI2LWVmMzEwMmFjZmY3NCI:1vDNl7:0iU6VJf4hAy92IyVMYrlWqG4gMhDG_h_DRy0el-6VKA" multiple="" data-select2-id="id_tags" tabindex="-1" aria-hidden="true">
    <option value="36" data-select2-id="2">Basket</option>
    <option value="35" data-select2-id="3">Team sport</option>
  </select>
  <span class="select2 select2-container select2-container--default select2-container--focus" dir="ltr" data-select2-id="1" style="width: 875px;">
    <span class="selection">
      <span class="select2-selection select2-selection--multiple" role="combobox" aria-haspopup="true" aria-expanded="false" tabindex="-1" aria-disabled="false">
        <ul class="select2-selection__rendered">
          <li class="select2-search select2-search--inline">
            <input class="select2-search__field" type="search" tabindex="0" autocomplete="off" autocorrect="off" autocapitalize="none" spellcheck="false" role="searchbox" aria-autocomplete="list" placeholder="Tags" style="width: 863px;">
          </li>
        </ul>
      </span>
    </span>
    <span class="dropdown-wrapper" aria-hidden="true"></span>
  </span>
  <span class="help-tooltip">
    A comma-separated list of tags.
  </span>
</label>
```

If you feel that I should create another thread, please let me know.

**Comment by codingjoe** [comment]
Hi @kunambi,

since this is a 6yo issue and the repo has since moved… yes, please open a new discussion :)

Best
Joe
